### DISCUSSION TITLE
spatie/laravel-permission (Cache user role)

### DISCUSSION THREAD

**Discussion by some0ne2** [spatie/laravel-permission#2678] [Q&A]
Dear laravel artisans,

I'm using the package `spatie/laravel-permission` for RBAC inside my application.
A middleware is running performing a simple role check:

    public function handle($request, Closure $next)
    {
        if (Auth::user()->hasRole('staff')) {
			//forward request...
        } else {
            abort(404);
        }
    }

Beside of this middleware I'm using a simple method inside a blade template file like:

    @if($user->canBeImpersonated())
    	//rendering content
    @endif

The `canBeImpersonated()` method from inside the `User` model is simple:

    //it's impossible to impersonate staff members
    public function canBeImpersonated()
    {
        return !$this->hasRole('staff');
    }

This will cause 2 SQL queries (doubled):

	select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` in (4) and `model_has_roles`.`model_type` = 'App\Models\User'
	select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` in (4) and `model_has_roles`.`model_type` = 'App\Models\User'

Is there an elegant way to prevent this from happening?

**Comment by erikn69** [comment]
My question would be: are `@if($user->canBeImpersonated())` and `Auth::user()` same model instance?? 
If it is the same record in two model instances, it will perform 2 queries, for each instance it has its own relationships and data
_even if it is the same record, same id_
